File: Searching & Sorting/search_in_rotated_sorted_array.cpp
class Solution {
public:
    int search(vector<int>& nums, int target) {
	    // // int f=0;
	    // for(int i=0;i<nums.size();i++)
	    // {
	    //     if(nums[i]==target)
	    //         return i;
	    // }
	    // return -1;
      
      
      *********************************************** O(lon n) **************************************** using binary search
        
        int s = 0;
        int e = nums.size()-1;
        int mid;
        
        while(s<=e){
            mid = (s+e)/2;
            if(nums[mid] == target){
                return mid;
            }
            else if(nums[mid] >= nums[s]){
                if(nums[s] <= target and target <= nums[mid]){
                    e = mid-1;
                }
                else{
                    s = mid+1;
                }
            }
            else{
                if(nums[mid] <= target and target <= nums[e]){
                    s = mid+1;
                }
                else{
                    e = mid-1;
                }
            }
            
        }
        return -1;
    }
};

File: GeeksforGeeks/Search in Rotated Sorted Array.java
class Solution {
    // int search(int[] arr, int key) {
        
    //     for(int i=0;i<arr.length;i++){
    //         if(arr[i]==key)
    //             return i;
    //     }
    //     return -1;
    // }
    
    // ****************** Above is O(n) solution, better is O(log(n)) ******
    
    
    public int findPivotIndex(int[] arr) {
        int s = 0, e = arr.length - 1;
        while (s <= e) {
            int mid = s + (e - s) / 2;
            if (mid < e && arr[mid] > arr[mid + 1]) return mid;
            if (mid > s && arr[mid - 1] > arr[mid]) return mid - 1;
            if (arr[s] <= arr[mid]) s = mid + 1;
            else e = mid - 1;
        }
        return -1;
    }

    public int binarySearch(int s, int e, int[] arr, int key) {
        while (s <= e) {
            int mid = s + (e - s) / 2;
            if (arr[mid] == key) return mid;
            if (arr[mid] < key) s = mid + 1;
            else e = mid - 1;
        }
        return -1;
    }

    public int search(int[] arr, int key) {
        int pivotIdx = findPivotIndex(arr);
        if (pivotIdx == -1) return binarySearch(0, arr.length - 1, arr, key);
        if (key >= arr[0] && key <= arr[pivotIdx]) {
            return binarySearch(0, pivotIdx, arr, key);
        }
        return binarySearch(pivotIdx + 1, arr.length - 1, arr, key);
    }
}
